Automattic/wp-calypso
# Make mixed atomic Jetpack e2e runs more retry-safe

Closes #82590 

## Proposed Changes

The overall goal of this PR is make our Jetpack Atomic Build Smoke able to handle retries. Previously, we were doing the ATOMIC_VARIATION randomization based on the jest runner ID. Now, we make a hash of the file name and use that. This way, we still get a nice mix of the variations, but when we do the automatic retries, we are guaranteed to use the same variation for the test that failed. This approach lets us bypass trying to store additional failure artifacts on the filesystem, which gets gross fast. 

To make this change, I had to refactor our `envVariables` file. Now, our ENV variables are accessed as getter properties on a class. This refactor is really nice for a couple of reasons!

First, this makes the transformation logic a lot cleaner: now each getter function just does whatever validation and transformation it needs. This reads cleaner and is easier to update in the future. (Think Open-Close principle.)

Also, most importantly, it allows you control _when_ those ENV values are evaluated. Previously, any import of this file would cause all ENV variables to be evaluated. That means if you needed part of the evaluation to be based on something (in this case, a global value from the Jest environment), you didn't have a way to include that! Now, you can defer that access.

To help with validation, I exposed a function called "validate" on the object that evaluates every property. We can now call this in a controlled way in our Jest Environment setup and still get all that good validation in one place!

## Testing Instructions

Run the Atomic Build Smoke build twice in a row. The same specs should use the same atomic variation!

## Pre-merge Checklist
